src/libostree/ostree-fetcher.c

```diff
@@ -937,20 +961,41 @@ on_request_sent (GObject        *object,
         }
       else if (!SOUP_STATUS_IS_SUCCESSFUL (msg->status_code))
         {
-          GIOErrorEnum code;
-          switch (msg->status_code)
+          pending->tried_mirrors++;
+
+          /* is there another mirror we can try? */
+          if (g_slist_next (pending->mirrorlist) != NULL)
             {
-            case 404:
-            case 410:
-              code = G_IO_ERROR_NOT_FOUND;
-              break;
-            default:
-              code = G_IO_ERROR_FAILED;
+              pending->mirrorlist = g_slist_next (pending->mirrorlist);
+              create_pending_soup_request (pending, &local_error);
+              g_queue_insert_sorted (&pending->thread_closure->pending_queue,
+                                     g_object_ref (task), pending_task_compare,
+                                     NULL);
+              remove_pending_rerun_queue (pending);
```

Comment:
We add to the pending queue, while `remove_pending_rerun_queue` removes from the outstanding queue since this is a new request. Maybe we should rename `remove_pending_rerun_queue` to something else. :abcd: